Changes in module 'ebi-metagenomics/samtools/bam2fq'
--- modules/ebi-metagenomics/samtools/bam2fq/main.nf
+++ modules/ebi-metagenomics/samtools/bam2fq/main.nf
@@ -21,7 +21,7 @@
 
     script:
     def args = task.ext.args ?: ''
-    def prefix = task.ext.prefix ?: "${meta.id}"
+    def prefix = task.ext.prefix ? "${task.ext.prefix}_${meta.id}": "${meta.id}"
 
     if (split) {
         """

************************************************************
