Changes in module 'nf-core/seqkit/grep'
--- modules/nf-core/seqkit/grep/main.nf
+++ modules/nf-core/seqkit/grep/main.nf
@@ -9,8 +9,7 @@
         'biocontainers/seqkit:2.4.0--h9ee0642_0' }"
 
     input:
-    tuple val(meta), path(sequence)
-    path pattern
+    tuple val(meta), path(sequence), path(pattern)
 
     output:
     tuple val(meta), path("*.{fa,fq}.gz")  , emit: filter

************************************************************
