Changes in module 'nf-core/megahit'
--- modules/nf-core/megahit/main.nf
+++ modules/nf-core/megahit/main.nf
@@ -33,6 +33,11 @@
             $args \\
             --out-prefix $prefix
 
+        if [ ! -s megahit_out/*.fa ]; then
+            echo "No contigs assembled" | tee /dev/stderr
+            exit 1
+        fi
+
         pigz \\
             --no-name \\
             -p $task.cpus \\
@@ -54,6 +59,11 @@
             $args \\
             --out-prefix $prefix
 
+        if [ ! -s megahit_out/*.fa ]; then
+            echo "No contigs assembled" | tee /dev/stderr
+            exit 1
+        fi
+
         pigz \\
             --no-name \\
             -p $task.cpus \\

************************************************************
