Changes in module 'nf-core/quast'
--- modules/nf-core/quast/main.nf
+++ modules/nf-core/quast/main.nf
@@ -29,7 +29,7 @@
     def features  = gff             ?  "--features $gff" : ''
     def reference = fasta           ?  "-r $fasta"       : ''
     """
-    quast.py \\
+    metaquast.py \\
         --output-dir $prefix \\
         $reference \\
         $features \\
@@ -37,10 +37,10 @@
         $args \\
         ${consensus.join(' ')}
 
-    ln -s ${prefix}/report.tsv ${prefix}.tsv
-    [ -f  ${prefix}/contigs_reports/all_alignments_transcriptome.tsv ] && ln -s ${prefix}/contigs_reports/all_alignments_transcriptome.tsv ${prefix}_transcriptome.tsv
-    [ -f  ${prefix}/contigs_reports/misassemblies_report.tsv         ] && ln -s ${prefix}/contigs_reports/misassemblies_report.tsv ${prefix}_misassemblies.tsv
-    [ -f  ${prefix}/contigs_reports/unaligned_report.tsv             ] && ln -s ${prefix}/contigs_reports/unaligned_report.tsv ${prefix}_unaligned.tsv
+    ln -s ${prefix}/combined_reference/report.tsv ${prefix}.tsv
+    [ -f  ${prefix}/combined_reference/contigs_reports/all_alignments_transcriptome.tsv ] && ln -s ${prefix}/combined_reference/contigs_reports/all_alignments_transcriptome.tsv ${prefix}_transcriptome.tsv
+    [ -f  ${prefix}/combined_reference/contigs_reports/misassemblies_report.tsv         ] && ln -s ${prefix}/combined_reference/contigs_reports/misassemblies_report.tsv ${prefix}_misassemblies.tsv
+    [ -f  ${prefix}/combined_reference/contigs_reports/unaligned_report.tsv             ] && ln -s ${prefix}/combined_reference/contigs_reports/unaligned_report.tsv ${prefix}_unaligned.tsv
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":
@@ -55,77 +55,77 @@
     def reference = fasta           ? "-r $fasta" : ''
 
     """
-    mkdir -p $prefix
-    touch $prefix/report.tsv
+    mkdir -p $prefix/combined_reference
     touch $prefix/report.html
-    touch $prefix/report.pdf
-    touch $prefix/quast.log
-    touch $prefix/transposed_report.txt
-    touch $prefix/transposed_report.tex
-    touch $prefix/icarus.html
-    touch $prefix/report.tex
-    touch $prefix/report.txt
+    touch $prefix/metaquast.log
+    touch $prefix/combined_reference/report.tsv
+    touch $prefix/combined_reference/report.pdf
+    touch $prefix/combined_reference/transposed_report.txt
+    touch $prefix/combined_reference/transposed_report.tex
+    touch $prefix/combined_reference/icarus.html
+    touch $prefix/combined_reference/report.tex
+    touch $prefix/combined_reference/report.txt
 
     mkdir -p $prefix/basic_stats
-    touch $prefix/basic_stats/cumulative_plot.pdf
-    touch $prefix/basic_stats/Nx_plot.pdf
-    touch $prefix/basic_stats/genome_GC_content_plot.pdf
-    touch $prefix/basic_stats/GC_content_plot.pdf
+    touch $prefix/combined_reference/basic_stats/cumulative_plot.pdf
+    touch $prefix/combined_reference/basic_stats/Nx_plot.pdf
+    touch $prefix/combined_reference/basic_stats/genome_GC_content_plot.pdf
+    touch $prefix/combined_reference/basic_stats/GC_content_plot.pdf
 
     mkdir -p $prefix/icarus_viewers
-    touch $prefix/icarus_viewers/contig_size_viewer.html
+    touch $prefix/combined_reference/icarus_viewers/contig_size_viewer.html
 
-    ln -s $prefix/report.tsv ${prefix}.tsv
+    ln -s $prefix/combined_reference/report.tsv ${prefix}.tsv
 
     if [ $fasta ]; then
-        touch $prefix/basic_stats/NGx_plot.pdf
-        touch $prefix/basic_stats/gc.icarus.txt
+        touch $prefix/combined_reference/basic_stats/NGx_plot.pdf
+        touch $prefix/combined_reference/basic_stats/gc.icarus.txt
 
-        mkdir -p $prefix/aligned_stats
-        touch $prefix/aligned_stats/NAx_plot.pdf
-        touch $prefix/aligned_stats/NGAx_plot.pdf
-        touch $prefix/aligned_stats/cumulative_plot.pdf
+        mkdir -p $prefix/combined_reference/aligned_stats
+        touch $prefix/combined_reference/aligned_stats/NAx_plot.pdf
+        touch $prefix/combined_reference/aligned_stats/NGAx_plot.pdf
+        touch $prefix/combined_reference/aligned_stats/cumulative_plot.pdf
 
-        mkdir -p $prefix/contigs_reports
-        touch $prefix/contigs_reports/all_alignments_transcriptome.tsv
-        touch $prefix/contigs_reports/contigs_report_transcriptome.mis_contigs.info
-        touch $prefix/contigs_reports/contigs_report_transcriptome.stderr
-        touch $prefix/contigs_reports/contigs_report_transcriptome.stdout
-        touch $prefix/contigs_reports/contigs_report_transcriptome.unaligned.info
-        mkdir -p $prefix/contigs_reports/minimap_output
-        touch $prefix/contigs_reports/minimap_output/transcriptome.coords
-        touch $prefix/contigs_reports/minimap_output/transcriptome.coords.filtered
-        touch $prefix/contigs_reports/minimap_output/transcriptome.coords_tmp
-        touch $prefix/contigs_reports/minimap_output/transcriptome.sf
-        touch $prefix/contigs_reports/minimap_output/transcriptome.unaligned
-        touch $prefix/contigs_reports/minimap_output/transcriptome.used_snps
-        touch $prefix/contigs_reports/misassemblies_frcurve_plot.pdf
-        touch $prefix/contigs_reports/misassemblies_plot.pdf
-        touch $prefix/contigs_reports/misassemblies_report.tex
-        touch $prefix/contigs_reports/misassemblies_report.tsv
-        touch $prefix/contigs_reports/misassemblies_report.txt
-        touch $prefix/contigs_reports/transcriptome.mis_contigs.fa
-        touch $prefix/contigs_reports/transposed_report_misassemblies.tex
-        touch $prefix/contigs_reports/transposed_report_misassemblies.tsv
-        touch $prefix/contigs_reports/transposed_report_misassemblies.txt
-        touch $prefix/contigs_reports/unaligned_report.tex
-        touch $prefix/contigs_reports/unaligned_report.tsv
-        touch $prefix/contigs_reports/unaligned_report.txt
+        mkdir -p $prefix/combined_reference/contigs_reports
+        touch $prefix/combined_reference/contigs_reports/all_alignments_transcriptome.tsv
+        touch $prefix/combined_reference/contigs_reports/contigs_report_transcriptome.mis_contigs.info
+        touch $prefix/combined_reference/contigs_reports/contigs_report_transcriptome.stderr
+        touch $prefix/combined_reference/contigs_reports/contigs_report_transcriptome.stdout
+        touch $prefix/combined_reference/contigs_reports/contigs_report_transcriptome.unaligned.info
+        mkdir -p $prefix/combined_reference/contigs_reports/minimap_output
+        touch $prefix/combined_reference/contigs_reports/minimap_output/transcriptome.coords
+        touch $prefix/combined_reference/contigs_reports/minimap_output/transcriptome.coords.filtered
+        touch $prefix/combined_reference/contigs_reports/minimap_output/transcriptome.coords_tmp
+        touch $prefix/combined_reference/contigs_reports/minimap_output/transcriptome.sf
+        touch $prefix/combined_reference/contigs_reports/minimap_output/transcriptome.unaligned
+        touch $prefix/combined_reference/contigs_reports/minimap_output/transcriptome.used_snps
+        touch $prefix/combined_reference/contigs_reports/misassemblies_frcurve_plot.pdf
+        touch $prefix/combined_reference/contigs_reports/misassemblies_plot.pdf
+        touch $prefix/combined_reference/contigs_reports/misassemblies_report.tex
+        touch $prefix/combined_reference/contigs_reports/misassemblies_report.tsv
+        touch $prefix/combined_reference/contigs_reports/misassemblies_report.txt
+        touch $prefix/combined_reference/contigs_reports/transcriptome.mis_contigs.fa
+        touch $prefix/combined_reference/contigs_reports/transposed_report_misassemblies.tex
+        touch $prefix/combined_reference/contigs_reports/transposed_report_misassemblies.tsv
+        touch $prefix/combined_reference/contigs_reports/transposed_report_misassemblies.txt
+        touch $prefix/combined_reference/contigs_reports/unaligned_report.tex
+        touch $prefix/combined_reference/contigs_reports/unaligned_report.tsv
+        touch $prefix/combined_reference/contigs_reports/unaligned_report.txt
 
-        mkdir -p $prefix/genome_stats
-        touch $prefix/genome_stats/genome_info.txt
-        touch $prefix/genome_stats/transcriptome_gaps.txt
-        touch $prefix/icarus_viewers/alignment_viewer.html
+        mkdir -p $prefix/combined_reference/genome_stats
+        touch $prefix/combined_reference/genome_stats/genome_info.txt
+        touch $prefix/combined_reference/genome_stats/transcriptome_gaps.txt
+        touch $prefix/combined_reference/icarus_viewers/alignment_viewer.html
 
-        ln -sf ${prefix}/contigs_reports/misassemblies_report.tsv ${prefix}_misassemblies.tsv
-        ln -sf ${prefix}/contigs_reports/unaligned_report.tsv ${prefix}_unaligned.tsv
-        ln -sf ${prefix}/contigs_reports/all_alignments_transcriptome.tsv ${prefix}_transcriptome.tsv
+        ln -sf ${prefix}/combined_reference/contigs_reports/misassemblies_report.tsv ${prefix}_misassemblies.tsv
+        ln -sf ${prefix}/combined_reference/contigs_reports/unaligned_report.tsv ${prefix}_unaligned.tsv
+        ln -sf ${prefix}/combined_reference/contigs_reports/all_alignments_transcriptome.tsv ${prefix}_transcriptome.tsv
 
     fi
 
     if ([ $fasta ] && [ $gff ]); then
-        touch $prefix/genome_stats/features_cumulative_plot.pdf
-        touch $prefix/genome_stats/features_frcurve_plot.pdf
+        touch $prefix/combined_reference/genome_stats/features_cumulative_plot.pdf
+        touch $prefix/combined_reference/genome_stats/features_frcurve_plot.pdf
     fi
 
     cat <<-END_VERSIONS > versions.yml

************************************************************
