Changes in module 'nf-core/quast'
--- modules/nf-core/quast/main.nf
+++ modules/nf-core/quast/main.nf
@@ -14,7 +14,7 @@
 
     output:
     tuple val(meta), path("${prefix}")                   , emit: results
-    tuple val(meta), path("${prefix}.tsv")               , emit: tsv
+    tuple val(meta), path("${prefix}.tsv")               , optional: true , emit: tsv
     tuple val(meta), path("${prefix}_transcriptome.tsv") , optional: true , emit: transcriptome
     tuple val(meta), path("${prefix}_misassemblies.tsv") , optional: true , emit: misassemblies
     tuple val(meta), path("${prefix}_unaligned.tsv")     , optional: true , emit: unaligned
@@ -29,7 +29,7 @@
     def features  = gff             ?  "--features $gff" : ''
     def reference = fasta           ?  "-r $fasta"       : ''
     """
-    quast.py \\
+    metaquast.py \\
         --output-dir $prefix \\
         $reference \\
         $features \\
@@ -55,11 +55,11 @@
     def reference = fasta           ? "-r $fasta" : ''
 
     """
-    mkdir -p $prefix
+    mkdir -p $prefix/combined_reference
+    touch $prefix/report.html
+    touch $prefix/metaquast.log
     touch $prefix/report.tsv
-    touch $prefix/report.html
     touch $prefix/report.pdf
-    touch $prefix/quast.log
     touch $prefix/transposed_report.txt
     touch $prefix/transposed_report.tex
     touch $prefix/icarus.html

************************************************************
